File: 1234-replace-the-substring-for-balanced-string/1234-replace-the-substring-for-balanced-string.cpp
class Solution {
public:
    /*
    
    Here, we have to find the length of the smallest substring that we have to remove such that the frequency
    of each character 'Q', 'W', 'E' and 'R' is less than or equal to n / 4
    
    We will use 2 pointers start and end to find this minimum size window that can be removed in order to get
    the balanced string
              0 1 2 3 4 5 6 7 8 9 10 11 
    I/P: s = "Q Q R E R E W E R Q  Q  E "
    O/P: 2
    Explanation: 
    freq < char , int >
         <   Q  ,  4  > 
         <   W  ,  1  >
         <   E  ,  4  >
         <   R  ,  3  >
    
    Remove QE from index 10-11 and replace them with WW 
    to get Q W Q W E R E W R Q W W where each character appears exactly n / 4 times
    
    */
    int balancedString(string s) {
        int n = s.length(), k = n / 4;
        
        // count frequency of occurence
        unordered_map<char, int> freq;
        for(char &c : s) freq[c]++;
        
        // sliding window logic
        int start = 0, end = 0, ans = n;
        for(end = 0; end < n; ++end) {
            // reduce the frequency
            freq[s[end]]--;
            // if it satisfies the condition, that frequency of each character <= n / 4
            // update our answer and move the start index to find the better window
            // start < n due to the corner case QWER or multiples of it
            while(start < n and freq['Q'] <= k and freq['W'] <= k and freq['E'] <= k and freq['R'] <= k) {
                ans = min(ans, end - start + 1);
                freq[s[start]]++;
                start++;
            }
        }
        return ans;
    }
};

File: 1234-replace-the-substring-for-balanced-string/1234-replace-the-substring-for-balanced-string.java
class Solution {
    public int balancedString(String s) {
        int n = s.length(), ans = n, k = n / 4;
        var freq = new HashMap<Character, Integer>();
        freq.put('Q', 0);
        freq.put('W', 0);
        freq.put('E', 0);
        freq.put('R', 0);
        
        for(int i = 0; i < n; ++i) {
            char c = s.charAt(i);
            freq.put(c, freq.get(c) + 1);
        }
        int i = 0;
        for(int j = 0; j < n; ++j) {
            char c = s.charAt(j);
            freq.put(c, freq.get(c) - 1);
            
            while(i < n && freq.get('Q') <= k && freq.get('W') <= k && freq.get('E') <= k && freq.get('R') <= k) {
                ans = Math.min(ans, j - i + 1);
                char ci = s.charAt(i);
                freq.put(ci, freq.get(ci) + 1);
                ++i;
            }
        }
        return ans;
    }
}